website/blog/2024/09-30-jsdom-based-scraping/index.md

```diff
@@ -0,0 +1,258 @@
+            apiValues[name] = value;
+        }
+        if (Object.values(apiValues).length === apiHeaderKeys.length) {
+            retries = 0;
+        }
+    };
+    window.XMLHttpRequest.prototype.open = (method, urlToOpen) => {
+        if (
+            ['static', 'scontent'].find((x) =>
+                urlToOpen.startsWith(`https://${x}`),
+            )
+        )
+        log.debug('urlToOpen', urlToOpen);
+    };
+    do {
+        await sleep(4000);
+        retries--;
+    } while (retries > 0);
+
+    await window.close();
+    return apiValues;
+};
+```
+
+In this function, we are creating a virtual console that uses `CustomResourceLoader` to run the background process and replace the browser with JSDOM.
+
+For this particular example, we need three mandatory headers to make the API call, and those are `anonymous-user-id,` `timestamp,` and `user-sign.`
+
+Using `XMLHttpRequest.prototype.setRequestHeader`, we are checking if the mentioned headers are in the response or not, if yeas, we take the value of those headers, and repeat the retries until we get all the headers.
+
+Then, the most important part is that we use `XMLHttpRequest.prototype.open` to extract the auth data and make calls without actually using browsers or exposing the bot activity.
+
+At the end of `createSessionFunction`, it returns a session with the required headers.
+
+Now coming to our main code, we will use CheerioCrawler and will use `prenavigationHooks` to inject the headers that we got from the earlier function into the `requestHandler`.
+
+```js
+const crawler = new CheerioCrawler({
+    sessionPoolOptions: {
+        maxPoolSize: 1,
+        createSessionFunction: async (sessionPool) =>
+            createSessionFunction(sessionPool, proxyConfiguration),
+    },
+    preNavigationHooks: [
+        (crawlingContext) => {
+            const { request, session } = crawlingContext;
+            request.headers = {
+                ...request.headers,
+                ...session.userData?.headers,
+            };
+        },
+    ],
+    proxyConfiguration,
+});
+```
+
+Finally in the request handler we make the call using the headers and make sure how many calls are needed to fetch all the data handling pagination.
+
+```js
+async requestHandler(context) {
+        const { log, request, json } = context;
+        const { userData } = request;
+        const { itemsCounter = 0, resultsLimit = 0 } = userData;
+        if (!json.data) {
+            throw new Error('BLOCKED');
+        }
+        const { data } = json;
+        const items = data.list;
+        const counter = itemsCounter + items.length;
+        const dataItems = items.slice(
+            0,
+            resultsLimit && counter > resultsLimit
+                ? resultsLimit - itemsCounter
+                : undefined,
+        );
+        await context.pushData(dataItems);
+        const {
+            pagination: { page, total },
+        } = data;
+        log.info(
```

Comment:
this is still misindented, 8 spaces instead of 4

Reply:
done 😸